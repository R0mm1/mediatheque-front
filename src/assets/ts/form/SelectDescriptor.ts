import AbstractDescriptor from "@/assets/ts/form/AbstractDescriptor";

export default class SelectDescriptor extends AbstractDescriptor {
    static defaultSearchable: boolean = false;

    descriptorType: string = 'SelectDescriptor';
    value: (undefined | string | null) = undefined;
    placeholder: string = '';
    options: { [index: string]: string };
    searchable: boolean = SelectDescriptor.defaultSearchable;

    constructor(name: string, options: { [index: string]: string }, label: string = '') {
        super(name);

        this.label = label;
        this.options = options;
    }

    getOptionsAsObjects(itemLabel: string = 'text', itemValue: string = 'value'): Object[] {
        let options: Object[] = [];
        Object.keys(this.options).forEach(optionName => {
            let option: any = {};
            option[itemLabel] = this.options[optionName];
            option[itemValue] = optionName;
            options.push(option);
        });
        return options;
    }

    setValue(value: string) {
        this.value = value;
        return this;
    }

    setPlaceholder(placeholder: string) {
        this.placeholder = placeholder;
        return this;
    }
}
