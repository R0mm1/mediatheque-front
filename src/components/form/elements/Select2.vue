<template>
    <FormElement container-custom-classes="form_element_select2" :no-label="selectDescriptor.noLabel">
        <template v-slot:element_content>
            <v-select :items="cOptions" :searchable="cIsSearchable" v-model="dataValue" :attach="element"
                      v-on:change="change"></v-select>
        </template>
    </FormElement>

</template>

<script lang="ts">
    import {Component, Emit, Prop, Vue} from "vue-property-decorator";
    import SelectDescriptor from "@/assets/ts/form/SelectDescriptor";
    import FormElement from "@/components/form/FormElement.vue";
    import LeftActionBarFormSelectDescriptor from "@/assets/ts/list/LeftActionBarFormSelectDescriptor";

    @Component({
        components: {FormElement}
    })
    export default class Select2 extends Vue {
        @Prop(Object) selectDescriptor!: SelectDescriptor;

        @Prop({type: Boolean, default: SelectDescriptor.defaultSearchable}) searchable!: boolean;
        @Prop({
            type: Object, default: () => {
            }
        }) options!: any;
        @Prop({type: String, default: ''}) value!: string;

        dataValue?: string;
        id?: number;
        element: (Node | null) = null;

        get cIsSearchable() {
            return this.hasDescriptorProvided ? this.selectDescriptor.searchable : this.searchable;
        }

        get cOptions() {
            return this.hasDescriptorProvided ? this.selectDescriptor.getOptionsAsObjects() : this.options;
        }

        get cValue() {
            return this.hasDescriptorProvided ? this.selectDescriptor.value : this.value;
        }

        get hasDescriptorProvided() {
            return this.selectDescriptor instanceof SelectDescriptor;
        }

        @Emit()
        change(value: any) {
            return value;
        }

        mounted() {
            this.element = this.$el.querySelector('.element_content');
        }

        created() {
            this.dataValue = this.cValue;
        }
    }
</script>

<style lang="scss">
    @import "../../../assets/scss/colors";

    .form_element_select2 {

        .element_content {
            position: relative;

            .v-menu__content, .v-select-list {
                border-radius: unset !important;
            }

            .v-input__slot {
                height: 2rem;
            }

            .v-menu__content {
                top: 0 !important;
            }

            .v-text-field {
                margin-top: 0 !important;
                padding-top: 0 !important;
            }

            .v-select__selections input {
                padding: 0 !important;
            }

            .v-select__selection--comma {
                margin: 0 !important;
            }

            .v-input__append-inner {
                margin-top: 0 !important;
            }

            .v-input__slot::before, .v-input__slot::after {
                border-style: none !important;
            }
        }
    }

</style>
