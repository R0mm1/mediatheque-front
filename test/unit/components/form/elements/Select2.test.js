import {createLocalVue, mount} from '@vue/test-utils'
import Select2 from "@/components/form/elements/Select2.vue"
import SelectDescriptor from "../../../../../src/assets/ts/form/SelectDescriptor";
import Vuetify from "vuetify";
import Vue from 'vue';

Vue.use(Vuetify);

describe('Select2', () => {

    const configBase = {
        created() {
            this.$vuetify.lang = {
                t: () => {
                },
            };
            this.$vuetify.theme = {dark: false};
        }
    };

    it('renders correctly', () => {
        const wrapper = mount(Select2, {
            // localVue: localVue,
            propsData: {
                'selectDescriptor': new SelectDescriptor('my-select', {
                    'option1': 'My first option',
                    'option2': 'My second option'
                }, 'This is an awesome select').setValue('option1'),
                'searchable': true,
                'value': 'option2'
            },
            ...configBase
        });
        expect(wrapper.element).toMatchSnapshot();
    });
});
