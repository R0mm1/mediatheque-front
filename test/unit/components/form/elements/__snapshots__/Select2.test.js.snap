// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select2 renders correctly 1`] = `
<div
  class="form_element form_element_select2"
>
  <label
    class="element_label"
    for=""
  >
    
            
        
  </label>
   
  <div
    class="element_content"
  >
    <div
      class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select"
    >
      <div
        class="v-input__control"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="list-3"
          class="v-input__slot"
          role="button"
        >
          <div
            class="v-select__slot"
          >
            <div
              class="v-select__selections"
            >
              <div
                class="v-select__selection v-select__selection--comma"
              >
                My first option
              </div>
              <input
                aria-readonly="false"
                autocomplete="off"
                id="input-3"
                readonly="readonly"
                type="text"
              />
            </div>
            <div
              class="v-input__append-inner"
            >
              <div
                class="v-input__icon v-input__icon--append"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate material-icons theme--light"
                >
                  $dropdown
                </i>
              </div>
            </div>
            <input
              type="hidden"
              value="option1"
            />
          </div>
          <div
            class="v-menu"
          >
            <!---->
          </div>
        </div>
        <div
          class="v-text-field__details"
        >
          <div
            class="v-messages theme--light"
          >
            <div
              class="v-messages__wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
